crates/provider/src/provider/get_block.rs

```diff
@@ -256,6 +257,74 @@ where
     }
 }
 
+/// A builder type for polling full blocks.
+///
+/// By default, this polls for blocks with [`BlockTransactionsKind::Hashes`].
+///
+/// The polling stream must be consumed by calling [`WatchBlocks::into_stream`].
+#[derive(Debug)]
+#[must_use = "this builder does nothing unless you call `.into_stream`"]
+pub struct WatchBlocks<BlockResp> {
+    poller: PollerBuilder<(U256,), Vec<B256>>,
+    kind: BlockTransactionsKind,
+    _pd: std::marker::PhantomData<BlockResp>,
+}
+
+impl<BlockResp> WatchBlocks<BlockResp>
+where
+    BlockResp: BlockResponse + RpcRecv,
+{
+    /// Create a new [`WatchBlocks`] instance.
+    pub fn new(poller: PollerBuilder<(U256,), Vec<B256>>) -> Self {
+        Self { poller, kind: BlockTransactionsKind::Hashes, _pd: PhantomData }
+    }
+
+    /// Poll for blocks with full transactions i.e [`BlockTransactionsKind::Full`].
+    pub fn full(mut self) -> Self {
+        self.kind = BlockTransactionsKind::Full;
+        self
+    }
+
+    /// Poll for blocks with just transactions hashes i.e [`BlockTransactionsKind::Hashes`].
+    pub fn hashes(mut self) -> Self {
+        self.kind = BlockTransactionsKind::Hashes;
+        self
+    }
+
+    /// Consumes the stream of block hashes from the inner [`PollerBuilder`] and maps it to a stream
+    /// of [`BlockResponse`].
+    pub fn into_stream(self) -> impl Stream<Item = TransportResult<BlockResp>> + Unpin {
```

Comment:
if we return anonymous types then this should also enforce Send because this can't be inferred on the callsite

but this is likely problematic for wasm compat

Reply:
we don't do this anywhere for streams

Reply:
you can call `into_stream` after sending, but yes ideally these are all named types...